### PR TITLE
feat(contracts): implement 1-step undelegation for KGH

### PR DESCRIPTION
**Pull request by 0xHansLee**
Implement 1-step undelegation for KGH (including batch)

### CODE REVIEW

**Comment by seolaoh** [comment]
Can you resolve compile errors in `AssetManager`?

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 1]
```diff
@@ -422,57 +437,87 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
         );
 
-        emit KghBatchUndelegationInitiated(validator, msg.sender, tokenIds, kroShares, kghShares);
+        // update validator tree
+        VALIDATOR_MANAGER.updateValidatorTree(validator, true);
+
+        // transfer KGH
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            KGH.safeTransferFrom(address(this), msg.sender, tokenIds[i]);
+        }
```
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:514]:
```suggestion
        for (uint256 i = 0; i < tokenIds.length; ) {
            KGH.safeTransferFrom(address(this), msg.sender, tokenIds[i]);
        
            unchecked {
                ++i;
            }
        }
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 2]
```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
         unchecked {
+            // asset
             vault.asset.totalKroShares -= kroShares;
-            vault.asset.totalKghShares -= kghShares;
-
-            uint128 baseRewardsToReceive = kroAssetsToWithdraw - kroInKghs;
-            boostedRewardsToReceive = kghAssetsToWithdraw - VKRO_PER_KGH * uint128(tokenIds.length);
-
-            vault.asset.totalKro -= kroAssetsToWithdraw;
+            vault.asset.totalKgh -= 1;
```
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:1029]:
```suggestion
            vault.asset.totalKgh -= kghCount;
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 3]
```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-                baseRewardsToReceive,
-                boostedRewardsToReceive,
-                pendingShares
-            );
-        }
 
-        if (kroAssetsToWithdraw + boostedRewardsToReceive > 0) {
-            VALIDATOR_MANAGER.updateValidatorTree(validator, true);
+            // delegator
+            kghDelegator.kghNum -= kghCount;
```
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:1034]:
```suggestion
            _vaults[validator].kghDelegators[delegator].kghNum -= kghCount;
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 4]
```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-        uint128 kroShares,
-        uint128 kghShares,
-        uint128 kghAssetsToWithdraw
+        uint128 kroAssets,
+        uint128 kroShares
     ) internal {
         Vault storage vault = _vaults[validator];
-        uint128 kroAssetsToWithdraw = _convertToKroAssets(validator, kroShares);
-        uint128 boostedRewardsToReceive;
+        KghDelegator storage kghDelegator = vault.kghDelegators[delegator];
```
Comment (sm-stack) [inline packages/contracts/contracts/L1/AssetManager.sol:1024]: How about just declaring `Asset storage asset`? That will be better for gas usage I think, combined with the comment at L#1034.

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 5]
```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
     /**
      * @notice Internal function to undelegate KGHs from the validator.
      *
      * @param validator           Address of the validator.
-     * @param tokenIds            Array of token ids of KGHs to undelegate.
+     * @param delegator           Address of the delegator.
+     * @param kghCount            The number of KGH token to undelegate.
      * @param kroInKghs           The amount of KRO in the KGHs.
+     * @param kroAssets           The amount of KRO Assets to undelegate.
      * @param kroShares           The amount of KRO shares to undelegate.
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:1013]: Please remove space

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 6]
```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
 
     /**
      * @notice Internal function to undelegate KGHs from the validator.
      *
      * @param validator           Address of the validator.
-     * @param tokenIds            Array of token ids of KGHs to undelegate.
+     * @param delegator           Address of the delegator.
+     * @param kghCount            The number of KGH token to undelegate.
      * @param kroInKghs           The amount of KRO in the KGHs.
+     * @param kroAssets           The amount of KRO Assets to undelegate.
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:1012]:
```suggestion
     * @param kroAssets           The amount of KRO assets to undelegate.
```

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 2 above)
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:1029]:
```suggestion
            vault.asset.totalKgh -= kghCount;
```

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 7]
```diff
@@ -340,7 +353,10 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
         address validator,
         uint256 tokenId
     ) external isRegistered(validator) returns (uint128) {
-        // TODO: claim boosted reward and delegate to KRO pool
+        // claim boost reward and compounding it
+        uint128 boostRewardAssets = _claimBoostedReward(validator, msg.sender);
+        uint128 boostRewardShares = _convertToKroShares(validator, boostRewardAssets);
+        _delegate(validator, msg.sender, boostRewardAssets, boostRewardShares);
```
[the remark this answers is not in the document]
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol:359]: I understood that KGH delegators are not stored in `valut.asset.kroDelegators`, but don't know why we need an internal function for auto-compounding. Could you elaborate more?

[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/AssetManager.sol:359]: We cannot use `_delegate` here, because it stores the information in `vault.asset.kroDelegators`

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 8]
```diff
@@ -500,17 +513,21 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
-            kroShares,
-            kroInKghs
+            kroShares
         );
 
         // update validator tree
         VALIDATOR_MANAGER.updateValidatorTree(validator, true);
 
         // transfer KGH
         for (uint256 i = 0; i < tokenIds.length; i++) {
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:525]: Please remove `i++`

**Review by seolaoh — APPROVED** [review]
LGTM

File: packages/contracts/contracts/L1/AssetManager.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xHansLee) [reply packages/contracts/contracts/L1/AssetManager.sol]: I see. As we discussed in slack, I edited to transfer the claimed boost reward when delegate KGH in [41ea83c](https://github.com/kroma-network/kroma/pull/353/commits/41ea83c9d3202a493eff446a559dbb7ab1d683af)

File: packages/contracts/contracts/L1/AssetManager.sol [hunk 9]
```diff
@@ -368,13 +368,9 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
...
     ) external isRegistered(validator) returns (uint128) {
-        // claim boost reward and auto compounding it
+        // claim boost reward
         uint128 boostRewardAssets = _claimBoostedReward(validator, msg.sender);
-        if (boostRewardAssets > 0) {
-            uint128 boostRewardShares = _convertToKroShares(validator, boostRewardAssets);
-            emit KghRewardClaimed(validator, msg.sender, boostRewardAssets, boostRewardShares);
-            _delegate(validator, msg.sender, boostRewardAssets, boostRewardShares);
-        }
+        ASSET_TOKEN.safeTransfer(msg.sender, boostRewardAssets);
```
Comment (seolaoh) [inline packages/contracts/contracts/L1/AssetManager.sol:373]: Why did you remove the if phrase?

**Review by sm-stack — APPROVED** [review]
LGTM
